chirp/drivers/kenwood_live.py

```diff
@@ -616,18 +1211,778 @@ class THD7GRadio(THD7Radio):
+    _TZ_OPTIONS = ["UTC - 12:00", "UTC - 11:30", "UTC - 11:00",
+                   "UTC - 10:30", "UTC - 10:00", "UTC - 9:30",
+                   "UTC - 9:00", "UTC - 8:30", "UTC - 8:00",
+                   "UTC - 7:30", "UTC - 7:00", "UTC - 6:30",
+                   "UTC - 6:00", "UTC - 5:30", "UTC - 5:00",
+                   "UTC - 4:30", "UTC - 4:00", "UTC - 3:30",
+                   "UTC - 3:00", "UTC - 2:30", "UTC - 2:00",
+                   "UTC - 1:30", "UTC - 1:00", "UTC - 0:30",
+                   "UTC", "UTC + 0:30", "UTC + 1:00", "UTC + 1:30",
+                   "UTC + 2:00", "UTC + 2:30", "UTC + 3:00",
+                   "UTC + 3:30", "UTC + 4:00", "UTC + 4:30",
+                   "UTC + 5:00", "UTC + 5:30", "UTC + 6:00",
+                   "UTC + 6:30", "UTC + 7:00", "UTC + 7:30",
+                   "UTC + 8:00", "UTC + 8:30", "UTC + 9:00",
+                   "UTC + 9:30", "UTC + 10:00", "UTC + 10:30",
+                   "UTC + 11:00", "UTC + 11:30", "UTC + 12:00"]
+
+    EXTRA_BOOL_SETTINGS = {
+        'aux': [("ELK", "Enable Tune When Locked"),
+                ("TH", "Tx Hold for 1750"),
+                ("TXS", "Transmit Inhibit")],
+        'dtmf': [("TXH", "TX Hold")],
+        'main': [("LMP", "Lamp")],
+    }
+    EXTRA_LIST_SETTINGS = {
+        'aprs': [("DSPA", "Display Area"),
+                 ("KILO", "Mile/Kilometer"),
+                 ("PAMB", "Position Ambiguity"),
+                 ("STXR", "Status Transmit Rate"),
+                 ("TEMP", "APRS Units"),
+                 ("TZ", "Timezone")],
+        'aux': [("CKEY", "CALL Key Set Up")],
+        'main': [("BAL", "Balance"),
+                 #("MNF", "Memory Display Mode")],  Only available in MR mode, not VFO mode
+                ],
+        'save': [("SV", "Battery Save")],
     }
 
-    def _get_setting_options(self, setting):
-        if setting in self._SETTINGS_OPTIONS:
-            opts = self._SETTINGS_OPTIONS[setting]
-        else:
-            opts = super()._SETTINGS_OPTIONS[setting]
-            if opts is None:
-                opts = getattr(self, '_%s_OPTIONS' % setting)
-        return opts
+    MAIDENHEAD_CHARSET = "0123456789ABCDEFGHIJKLMNOPQRSTUVWX" + \
+                         "abcdefghijklmnopqrstuvwx"
+    EXTRA_STRING_SETTINGS = {
+        'aprs': [("ARLM", "Auto Reply Message", 45),
+                 ("AMGG", "Message Groups", 45, "ABCDEFGHIJKLMNOPQRSTUVWXYZ" +
+                                                "*,-0123456789"),
+                 ("MP 1", "My Position 1", 10, MAIDENHEAD_CHARSET),
+                 ("MP 2", "My Position 2", 10, MAIDENHEAD_CHARSET),
+                 ("MP 3", "My Position 3", 10, MAIDENHEAD_CHARSET)],
+    }
+
+    EXTRA_MAP_SETTINGS = {
+        'main': [("BEL 0", "Tone Alert Band A"),
+                 ("BEL 1", "Tone Alert Band B"),
+                 ("SQ 0", "Band A Squelch"),
+                 ("SQ 1", "Band B Squelch")],
+        'aprs': [("TXD", "Transmit Delay")],
+    }
```

Comment:
I wonder if it makes sense to pull the D7 out to another file at this point? You've added a lot to this file, which was already too big. You can just import `kenwood_live` and do the same thing, but keep all this verbose settings stuff to one file. What do you think? If you do, it'd be great if you could make one commit that splits the existing driver to a new file (unchanged) and then another commit to make your changes here, just so it's still easy to review the delta. I know that's more work, but hopefully you'll agree it's worthwhile. This file has just grown organically beyond what it really should have in it :)

Maybe move the D700 in there as well?

Reply:
Makes sense, I was already considering it... at the same time, I'll make an abstract base class that they all share and hopefully come up with something better than the _SETTINGS_OPTIONS{}/_XXX_OPTIONS override mechanism that's currently in use.

The maidenhead conversion stuff doesn't *really* belong here either, it was just the easiest way to get lat/lon data into a single text box.  The radio is set in XX°YY.YY′ format, but having four boxes all controlling the same set command felt more awkward to me.

Reply:
Okay, cool. This definitely needs some cleanup, we just need to be careful not to disturb all the drivers under this that will be hard to find testers for. I have a D700, D710G, D72, D74, and TM271 I can use for verification and I *think* @KC9HI has an F6A, IIRC.